File: 1631-number-of-sub-arrays-with-odd-sum/1631-number-of-sub-arrays-with-odd-sum.go
func numOfSubarrays(arr []int) int {
    const MOD = 1e9 + 7
    
    prefixSum, oddCount, evenCount := 0, 0, 1
    oddSubarraysCount := 0

    for _, val := range arr {
        prefixSum += val

        if (prefixSum & 1) == 1 {
            oddSubarraysCount += evenCount
            oddCount++
        } else {
            oddSubarraysCount += oddCount
            evenCount++
        }

        oddSubarraysCount %= MOD
    }

    return oddSubarraysCount
}

File: 1631-number-of-sub-arrays-with-odd-sum/1631-number-of-sub-arrays-with-odd-sum.cpp
class Solution {
public:
    int numOfSubarrays(vector<int>& arr) {
        const int MOD = 1e9 + 7;
        int n = arr.size();
        int oddCount = 0, evenCount = 1, prefixSum = 0;
        int oddSubarraysCount = 0;

        for (int i = 0; i < n; i++) {
            prefixSum += arr[i];

            if (prefixSum & 1) {
                oddSubarraysCount += evenCount;
                oddCount++;
            } else {
                oddSubarraysCount += oddCount;
                evenCount++;
            }

            oddSubarraysCount %= MOD;
        }

        return oddSubarraysCount;
    }
};